resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/sse/SseEventOutputImpl.java

```diff
@@ -182,25 +197,58 @@ private void internalFlushResponseToClient(boolean throwIOException) throws IOEx
 
             try
             {
+               CompletableFuture<Void> ret = new CompletableFuture<>();
                ServerResponseWriter.writeNomapResponse(jaxrsResponse, request, response,
                      ResteasyProviderFactory.getInstance(), t -> {
+                        AsyncOutputStream aos;
+                        try
+                        {
+                           aos = response.getAsyncOutputStream();
+                        } catch (IOException x)
+                        {
+                           close(false);
+                           ret.completeExceptionally(x);
+                           return;
+                        }
+                        // eager composition to guarantee ordering
```

Comment:
I don't understand this comment? Ordering is important everywhere and nowhere else seems to use this pattern?

Reply:
That's special to SSE, which supports CS in the API as a return type, but the JAX-RS spec always gives examples where they call `CS<> writeEvent()` in sequence without chaining them and using their return type. This is super wrong, because if you use `compose` in `writeEvent` then it's possible only the first part of the composition is actually called by the time `writeEvent` returns.

So for example, if `writeEvent` is a composition of `A, B`, and you call `writeEvent(); writeEvent()` it's very possible that the sequence of calls end up being `A, A, B, B`.

This only happens if you ignore the return value of `writeEvent` which is WRONG, but since the JAX-RS spec examples do that, I've had to support the existing tests doing it too.

We can't support this sort of behaviour everywhere though, so for the new async io APIs I don't want to support this broken pattern, and we want users to actually use the returned CS for proper chaining, which is why I haven't supported this behaviour elsewhere.

Reply:
ok, makes sense